# Assignment 1

Students: Clara Louise Brodt (s224189) & Celina Laungaard (s234809)

Course: Social Computational Science (02467)

GitHub link: 

## Part 1 - Web-scraping

In [37]:
from bs4 import BeautifulSoup # A package to work with HTML data
import requests #A package to make HTTP requests
import unicodedata
import re # Used to remove single letters in names
import pandas as pd

In [38]:
LINK = "https://ic2s2-2023.org/program" # Get link

r = requests.get(LINK) # Use Requests package

soup = BeautifulSoup(r.content) # Soup it

In [39]:
names = soup.find_all("i")
names

[<i>Chair: Claudia Wagner</i>,
 <i><u>Jonas L Juul</u>, Jon Kleinberg</i>,
 <i><u>Chloe Ahn</u>, Xinyi Wang</i>,
 <i><u>Giuseppe Russo</u>, luca verginer, Manoel Horta Ribeiro, Giona Casiraghi</i>,
 <i><u>Almog Simchon</u>, Adam Sutton, Matthew Edwards, Stephan Lewandowsky</i>,
 <i><u>Arianna Pera</u>, Manuel Vimercati, Matteo Palmonari</i>,
 <i><u>Mohammed Alsobay</u>, Abdullah Almaatouq, David G. Rand, Duncan J. Watts</i>,
 <i>Sara Venturini, <u>Satyaki Sikdar</u>, Francesco Rinaldi, Francesco Tudisco, Santo Fortunato</i>,
 <i><u>Isabella Loaiza</u>, Takahiro Yabe, Alex Pentland</i>,
 <i>Chair: Taha Yasseri</i>,
 <i><u>Silvia De Sojo Caso</u>, Mia Ann Jørgensen, Sune Lehmann, Laura Alessandretti</i>,
 <i><u>Emil Bakkensen Johansen</u>, Mathias Wullum Nielsen</i>,
 <i><u>Rubén Rodríguez Casañ</u>, Antonio Ariño Villarroya</i>,
 <i><u>Sunny Rai</u>, Ashley Francisco, Salvatore Giorgi, Brenda Curtis, Lyle Ungar, Sharath Chandra Guntuku</i>,
 <i><u>Allison Koenecke</u>, Eric Giannella, R

In [40]:
# Assuming 'names' is a ResultSet (list of elements) returned by find_all()
clean_names = [name.text for name in names]  # Extract text from each element

# Join all names into a single string and split on ',' to handle multiple names
clean_names = [name.strip() for name in ", ".join(clean_names).split(",")]

#Remove "Chair: " (case-insensitive) and strip whitespace
clean_names = [name.replace("Chair: ", "").replace("chair: ", "").strip() for name in clean_names]

#Remove duplicates while maintaining order
seen = set()
clean_names = [name for name in clean_names if not (name in seen or seen.add(name))]


clean_names = [name.strip() for name in ", ".join(clean_names).split(",")]  # Normalize spacing and split names
clean_names = ["".join(c for c in unicodedata.normalize("NFD", name) if not unicodedata.combining(c)) for name in clean_names]  # Remove diacritics (hats)
clean_names = [name.replace("-", " ").replace("."," ") for name in clean_names]  # Replace - with space
clean_names = [re.sub(r'\s+[A-Za-z]\s+', ' ', name) for name in clean_names] # Removes single letters in names

clean_names = set(clean_names) # Remove all exact duplicates

print(clean_names)  # List of cleaned unique names

len(clean_names)

{'Shima Saniei', 'Luisa Carrer', 'Jake Hofman', 'Lauren Brent', 'Patrick Gildersleve', 'Frank van Tubergen', 'Natalie McDaniel Gallagher', 'Christiern Santos Rasmussen', 'Minsu Park', 'Pietro Alessandro Aluffi', 'Kangsan Lee', 'David Lagnado', 'Rupert Tibor Kiddle', 'Sean Rogers', 'Hal Ashton', 'Katie Spoon', 'George Bikhazi', 'Denny Borsboom', 'Bramantyo Supriyatno', 'Rada Mihalcea', 'Yevgeniy Golovchenko', 'Sune Lehmann', 'Sarah Ita Levitan', 'Regina Widjaya', 'Holger Steinmetz', 'Ting Lian', 'Ryan Cotterell', 'Federico Delussu', 'Barbara Keller', 'Taro Takaguchi', 'Mihai Burzo', 'Marton Rakovics', 'Mael Kubli', 'Kobi Hackenburg', 'Iyad Rahwan', 'Ichiro Sakata', 'Marie Laure Charpignon', 'Almog Simchon', 'Mingjing Yin', 'Eric Dignum', 'Mohammad Atari', 'Scott Hale', 'Anya Hommadova Lu', 'Tomomi Kito', 'Taberez Ahmed Neyazi', 'Eddie Rossi', 'Abdullah Almaatouq', 'Francesco Tudisco', 'Neville Stanton', 'Cecilia Monge', 'Coen Needell', 'Alexandra Rottenkolber', 'Lorenzo Amir Nemati Fard

1459

In [42]:
df = pd.DataFrame(clean_names) # Creates pandas dataframe

df.to_csv("Authors_2023.csv",index=False) # Save in seperate file

### Part 1.5
**How many unique researchers do you get?**


In total we got 1459 unique researchers.

### Part 1.6

**Explain the process you followed to web-scrape the page. Which choices did you make to accurately retreive as many names as possible? Which strategies did you use to assess the quality of your final list? Explain your reasoning and your choices (answer in max 150 words).**

We used the BeautifulSoup package to collect the raw data from the program. After examining the webpage, we discovered that all the names were written in italics, so we collected all italicized words. After removing noise, such as \<i>, \<u>, and "Chair: ", we obtained a list of names. Finally, we replaced all dots, commas, single letters in names, and removed duplicates.

# Part 2 - Ready Made vs Custom Made Data

### Part 2.1
**What are pros and cons of the custom-made data used in Centola's experiment (the first study presented in the lecture) and the ready-made data used in Nicolaides's study (the second study presented in the lecture)? You can support your arguments based on the content of the lecture and the information you read in Chapter 2.3 of the book __(answer in max 150 words)__.**


Nicolaides’ study relies on ready-made data, which has the pro of being large, continuously available, and non-reactive, making it ideal for detecting small differences. However, a con is that this data is often inaccessible, non-representative, and contains many confounding factors—some accounted for, while others remain unaddressed. In contrast, Centola’s study uses custom-made data, with the pros of being cleaner, more reproducible, and containing fewer confounding elements, enhancing its reliability. A con of Centola’s approach is the smaller scale of the experiment, which may oversimplify social networks, making it difficult to fully capture real-world complexity.

### Part 2.2
**How do you think these differences can influence the interpretation of the results in each study? __(answer in max 150 words)__**

When interpreting Nicolaides' study, it is possible to detect small but significant patterns in people's behavior due to the large-scale data. However, one must be aware that the data might not be representative, as not all elements are controlled, allowing the possibility of bias. It can therefore be difficult to precisely determine from where and why behavior occurs.

Centola's study allows for control over these elements and can provide strong causal evidence of behaviors. However, since it is collected in a controlled environment, it may not accurately reflect real-world scenarios, making it challenging to generalize results from the lab to real life.

# Part 3 - Gathering Research Articles using the OpenAlex API

### Retrieve Data

In [ ]:
import requests
import seaborn as sns
import pandas as pd
from tqdm import tqdm
from time import sleep

In [5]:
#filtrer authors to include only those > 5
all_authors = pd.read_csv("author_data2024.csv", header=None) #all authors
all_authors[2] = pd.to_numeric(all_authors[2], errors="coerce") #from string to number

filtered_authors = all_authors[all_authors[2] > 5]  #Remove all below 5
filtered_authors = filtered_authors[filtered_authors[2] < 5000] #Remove all above 5000

df = filtered_authors[0]
df

1       https://openalex.org/A5105369812
2       https://openalex.org/A5100329996
3       https://openalex.org/A5100672274
5       https://openalex.org/A5110117708
6       https://openalex.org/A5010287399
                      ...               
1199    https://openalex.org/A5032167625
1201    https://openalex.org/A5033574079
1202    https://openalex.org/A5100766043
1204    https://openalex.org/A5075885830
1205    https://openalex.org/A5108220394
Name: 0, Length: 965, dtype: object

In [ ]:
import requests
import pandas as pd
from time import sleep
from tqdm import tqdm
from joblib import Parallel, delayed

BASE_URL = "https://api.openalex.org/works?filter="

# Define a function to fetch data for a single author
def fetch_work_data(author):
    extracted_data = []
    Filter1 = "author.id:"
    Filter2 = ",authors_count:<10,cited_by_count:>10,concepts.id:C126322002|C136264566|C19165224|C17744445,concepts.id:C33923547|C2522767161|C41008148"
    Filter3 = "&per-page=200&select=id,title,publication_year,cited_by_count,concepts,display_name,abstract_inverted_index,authorships&cursor="
    cursor = "*"
    
    while cursor:
        URL = BASE_URL + Filter1 + author + Filter2 + Filter3 + cursor
        
        response = requests.get(URL)
        
        if not response.ok:
            print(f"Error fetching data for {author}: {response.status_code}")
            sleep(1)
            continue
        
        json_data = response.json()
        cursor = json_data["meta"].get("next_cursor")
        
        for work in json_data.get("results", []):
            extracted_data.append({
                "id": work.get("id", ""),
                "publication_year": work.get("publication_year", ""),
                "cited_by_count": work.get("cited_by_count", 0),
                "author_ids": [auth["author"]["id"] for auth in work.get("authorships", [])],
                "title": work.get("title", ""),
                "abstract_inverted_index": work.get("abstract_inverted_index", None)
            })
    
    return extracted_data

# Run parallel processing
num_jobs = -1  
author_list = df 
all_extracted_data = Parallel(n_jobs=num_jobs)(delayed(fetch_work_data)(author) for author in tqdm(df))

# Flatten list of lists
final_data = [item for sublist in all_extracted_data for item in sublist]

# Convert to DataFrame
extracted_data = pd.DataFrame(final_data)

### Data storage


In [ ]:
# Convert dictionary to DataFrame
works_df = pd.DataFrame(extracted_data)

works_df.to_csv('Works_by_2024_authors', index=False)

In [ ]:
works_df = pd.read_csv("Works_by_2024_authors.csv")

# Select specific columns for papers_data
papers_data = works_df[[ "id", "publication_year", "cited_by_count", "author_ids" ]]

# Select specific columns for abstract_data
abstract_data = works_df[[ "id", "title", "abstract_inverted_index" ]]

In [ ]:
# Save filtered dataFrames to CSV
papers_data.to_csv("papers_data.csv", index=False)
abstract_data.to_csv("abstract_data.csv", index=False)

### Data Overview and Reflection questions: 

> - **Dataset summary.** How many works are listed in your *IC2S2 papers* dataframe? How many unique researchers have co-authored these works? 

In total there are 5049 works. 

In [ ]:
works_df = pd.read_csv("Works_by_2024_authors.csv")
full_author_list = works_df["author_ids"]

list_of_coauthors = []

#Create one list of all authors
for list in tqdm(full_author_list):
    for author in list:
        list_of_coauthors.append(author)

#Remove duplicates of authors 
unique_coauthors = []
urls = []
for author in list_of_coauthors: 
    if author not in unique_coauthors: 
        number = author.split('A')[-1]
        unique_coauthors.append(author)
        urls.append(number)

#Print lenght of total number of authors and total number of unique authors
len(unique_coauthors),len(list_of_coauthors)

In total there are 7350 unique co-authors.

> - **Efficiency in code.** Describe the strategies you implemented to make your code more efficient. How did your approach affect your code's execution time? __(answer in max 150 words)__



At first we only filtered the works directly in our request and pulled 200 works at a time to get the most optimal running time. Here we fetched all the works by a single author without running anything parallel and it resulted in it taking more than 20 minutes to run the code and extract the data. 
After implementing the suggested Parallel function from Joblib it reduced the running time down to around 3 minutes, which was a significant reduction. 
We ran into the issue of us being denied access through the API which would ruin the extracted data as the code would then skip this work. To avoid this we introduced an if statement to check our response and if it was not succesful, we would sleep and redo that request to ensure all works where collected. 

> - **Filtering Criteria and Dataset Relevance** Reflect on the rationale behind setting specific thresholds for the total number of works by an author, the citation count, the number of authors per work, and the relevance of works to specific fields. How do these filtering criteria contribute to the relevance of the dataset you compiled? Do you believe any aspects of Computational Social Science research might be underrepresented or overrepresented as a result of these choices? __(answer in max 150 words)__

By setting specific thresholds, the dataset focuses on influential works, highlighting key contributions. These filtering criteria remove less impactful studies, ensuring the dataset remains relevant. However, this might cause emerging authors to be underrepresented, while more established ones could be overrepresented. Additionally, some collaborative fields might be unevenly distributed if many excluded works come from those areas.

# Part 4 - The Network of Computational Social Science

### Part 1 - Network Construction

In [ ]:
import pandas as pd
from itertools import combinations
from collections import Counter

In [ ]:
## 

co_authors = pd.read_csv("papers2024co_authors.csv")
authors = pd.read_csv("papers2024authors.csv")

works = pd.concat([authors,co_authors],ignore_index = True)

#Create DataFrame
df = pd.DataFrame(works, columns=['id', 'publication_year', 'cited_by_count', 'author_ids'])

# Convert author_ids from string to list
df['author_ids'] = df['author_ids'].apply(eval)

# Generate all author pairs and count collaborations
collab_counter = Counter()

for authors in df['author_ids']:
    for pair in combinations(sorted(authors), 2):  # Sort to avoid duplicates (A, B) and (B, A)
        collab_counter[pair] += 1

# Convert Counter to DataFrame
collab_df = pd.DataFrame(collab_counter.items(), columns=['author_pair', 'collaboration_count'])

# Split tuples into separate columns for clarity
collab_df[['author_1', 'author_2']] = pd.DataFrame(collab_df['author_pair'].tolist(), index=collab_df.index)
collab_df.drop(columns=['author_pair'], inplace=True)
collab_df

In [ ]:
import networkx as nx
import matplotlib.pyplot as plt

#Load datasets
co_authors = pd.read_csv("co_author_data.csv")
authors = pd.read_csv("author_data.csv")

co_authors_papers = pd.read_csv("papers2024co_authors.csv")
authors_papers = pd.read_csv("papers2024authors.csv")

#Combine co_author and author datasets
papers = pd.concat([authors_papers,co_authors_papers],ignore_index = True)
author_data = pd.concat([authors,co_authors],ignore_index = True)

#Create undirected graph
G = nx.Graph()

#Compute total citations per author
papers['author_ids'] = papers['author_ids'].apply(eval)  # Convert from string to list if needed
papers_exploded = papers.explode('author_ids')
author_citations = papers_exploded.groupby('author_ids')['cited_by_count'].sum().to_dict()

#Compute first publication year for each author
author_years = papers.explode('author_ids')[['author_ids', 'publication_year']]
first_pub_years = author_years.groupby('author_ids')['publication_year'].min().todict()

#Add nodes (authors) with attributes
for _, row in author_data.iterrows():
    G.add_node(row["id"], 
               display_name=row['display_name'], 
               country=row.get('country_code', 'Unknown'), 
               first_pub_year=first_pub_years.get(row["id"], "Unknown"), 
               total_citations=author_citations.get(row["id"], 0))

#Add weighted edges (collaborations)
edges = [(row['author_1'], row['author_2'], row['collaborationcount']) for _, row in collab_df.iterrows()]
G.add_weighted_edges_from(edges)


In [ ]:
import json

#Save graph as JSON
graph_data = nx.node_link_data(G)
with open("author_network.json", "w") as f:
    json.dump(graph_data, f, indent=4)


### Part 2: Preliminary Network Analysis
> Now, with the network constructed, perform a basic analysis to explore its features.
> 1. **Network Metrics:**
>    - What is the total number of nodes (authors) and links (collaborations) in the network? 
>    - Calculate the network's density (the ratio of actual links to the maximum possible number of links). Would you say that the network is sparse? Justify your answer.
>    - Is the network fully connected (i.e., is there a direct or indirect path between every pair of nodes within the network), or is it disconnected?
>    - If the network is disconnected, how many connected components does it have? A connected component is defined as a subset of nodes within the network where a path exists between any pair of nodes in that subset. 
>    - How many isolated nodes are there in your network?  An isolated node is defined as a node with no connections to any other node in the network.
>    - Discuss the results above on network density, and connectivity. Are your findings in line with what you expected? Why?  __(answer in max 150 words)__
> 

In [ ]:
import json

# Load the JSON file
with open("author_network.json", "r") as f:
    data = json.load(f)

# Count nodes and links
num_nodes = len(data["nodes"])  # Nodes are usually stored in 'nodes' key
num_edges = len(data["links"])  # Links (edges) are stored in 'links' key

print(f"Number of nodes: {num_nodes}")
print(f"Number of links: {num_edges}")

In total 61580 nodes and 204125 links. 

> 2. **Degree Analysis:**
>    - Compute the average, median, mode, minimum, and maximum degree of the nodes. Perform the same analysis for node strength (weighted degree). What do these metrics tell us about the network? __(answer in max 150 words)__



In [ ]:
import statistics

# Compute degree (number of connections per node)
degrees = [deg for _, deg in G.degree()]
weighted_degrees = [deg for _, deg in G.degree(weight='weight')]

# Compute statistics
degree_stats = {
    "average_degree": sum(degrees) / len(degrees),
    "median_degree": statistics.median(degrees),
    "mode_degree": statistics.mode(degrees) if len(set(degrees)) > 1 else "No mode",
    "min_degree": min(degrees),
    "max_degree": max(degrees)
}

weighted_degree_stats = {
    "average_weighted_degree": sum(weighted_degrees) / len(weighted_degrees),
    "median_weighted_degree": statistics.median(weighted_degrees),
    "mode_weighted_degree": statistics.mode(weighted_degrees) if len(set(weighted_degrees)) > 1 else "No mode",
    "min_weighted_degree": min(weighted_degrees),
    "max_weighted_degree": max(weighted_degrees)
}

# Print results
print("Degree Statistics:", degree_stats)
print("Weighted Degree Statistics:", weighted_degree_stats)

The average degree of 6.63 suggests moderate collaboration, while the median of 5.0 and mode of 4 indicates that most of the authors have a small number of co-authors. However, the maximum degree of 420 reveals that a few of the authors is highly connected and probably central figures in the field. The weighted degree follows a similar pattern, with an average of 14.28 but a maximum of 1517, showing again that some of the authors have a really high collaboration intensity. This analysis shows a few highly influential authors bridge different research clusters, while most authors collaborate within smaller groups.

> 
> 3. **Top Authors:**
>    - Identify the top 5 authors by degree. What role do these node play in the network? 
>    - Research these authors online. What areas do they specialize in? Do you think that their work aligns with the themes of Computational Social Science? If not, what could be possible reasons? __(answer in max 150 words)__

In [ ]:
#Compute degrees for all nodes
degree_dict = dict(G.degree())

#Get the top 5 authors by degree (most connections)
top_5_authors = sorted(degree_dict.items(), key=lambda x: x[1], reverse=True)[:5]

#Print results
print("Top 5 Authors by Degree:")
for author, degree in top_5_authors:
    print(f"Author: {author}, Degree: {degree}")

The top 5 authors by degree are the most well-connected individuals in the network. These authors act as center points in network as they collaborate with a large number of other authors. Their high degree  suggests that they play a crucial role in connecting different research groups, and that they are also contributing to a lot of works and therefor have a lot of knowlegde. The 5 authors are David Moher, Ara Darzi, Shuicheng Yan, Philip s. Yu and Witold Pedrycz. They are all working at different universities or colleges from around the world. Moher,Yan, Yu all specialize in areas which fits well with computational science such as Meta-analysis and Video retrieval. Ara Darzi specialize in topics related more to the medical scene but with its focus on research and techniques it also aligns with the themes. Lastly Pedrucz specialize in machine learning and decision making, which seems the most different from the themes, but as they fit under the computer science theme it makes sense that he is included. 